### PR TITLE
[RFC] cql-pytest: run all tests for both vnodes and tablets

### PR DESCRIPTION
**Pull request by denesb**
This PR is an RFC, opened to start a discussions around how to get cql-pytest to run with both tablets and vnodes. Going forward, both will be supported, at least for some time.
This PR:
* Adds a new command-line option to pytest: `--replication={vnodes,tablets}`
* Modifies run to:
    - enable consistent topology changes and tablets when starting scylla
    - invoke pytest twice: once with `--replication=vnodes` and once with `--replication=tablets`
* Modifies `test_keyspace` to look for the new command line option and enable tablets if needed
* Provide two new fixtures: `vnodes_only` and `tablets_only` so tests that only work with on or the other can opt out. We will probably mainly use this at first to get all tests passing, annotating skipped tests with an issue number.

### CODE REVIEW

**Comment by denesb** [comment]
As expected running the tests with tablets fails. What is less expected, is that running with tablets takes twice as long:
Vnodes:
```
===== 1094 passed, 33 skipped, 216 xfailed, 6 xpassed in 272.65s (0:04:32) =====                                                                                                                                                                        
```

Tablets:
```
= 13 failed, 1081 passed, 33 skipped, 216 xfailed, 6 xpassed in 535.79s (0:08:55) =                                                                                                                                                                     
```

File: test/cql-pytest/conftest.py [hunk 1]
```diff
@@ -93,12 +95,30 @@ def this_dc(cql):
...
-    cql.execute("CREATE KEYSPACE " + name + " WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 1 }")
+    replication_options = [("class", "NetworkTopologyStrategy"), (this_dc, 1)]
+    if request.config.getoption("replication") == "tablets":
+        replication_options.append(("initial_tablets", 128))
+    serialized_options = ", ".join([f"'{key}' : '{value}'" for key, value in replication_options])
+    cql.execute("CREATE KEYSPACE " + name + " WITH REPLICATION = { " + serialized_options + " }")
     yield name
     cql.execute("DROP KEYSPACE " + name)
 
+# For tests that don't work with tablets (yet).
```
Comment (nyh) [inline test/cql-pytest/conftest.py:108]: As I commented elsewhere, I think this approach is wrong. We don't have "tests that don't work with tablets" - we have **code** that doesn't work with tablets, causing the test to fail. So we shouldn't have something called "vnode_only", we should have "xfail_on_tablets".

We may indeed need a "tablets_only" for tests that are designed directly to test the tablets feature, but note that I create a couple of days ago (I'm not sure if it went in already?) test/cql-pytest/test_tablets.py, and this file already had its own skip-if-not-tablets function so we don't necessarily need aglobal one in conftest.py

File: test/cql-pytest/conftest.py [hunk 2]
```diff
@@ -93,12 +95,30 @@ def this_dc(cql):
 # and automatically deleted at the end. We use scope="session" so that all
 # tests will reuse the same keyspace.
 @pytest.fixture(scope="session")
-def test_keyspace(cql, this_dc):
+def test_keyspace(request, cql, this_dc):
     name = unique_name()
-    cql.execute("CREATE KEYSPACE " + name + " WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 1 }")
+    replication_options = [("class", "NetworkTopologyStrategy"), (this_dc, 1)]
+    if request.config.getoption("replication") == "tablets":
+        replication_options.append(("initial_tablets", 128))
```
Comment (nyh) [inline test/cql-pytest/conftest.py:102]: I don't think 128 is a good default. I'd use 4. This may explain why your tests became very slow - it's a lot of work to create 128 tablets.

File: test/cql-pytest/run [hunk 3]
```diff
@@ -27,9 +27,14 @@ run.wait_for_services(pid, [
     lambda: run.check_rest_api(ip),
     lambda: check_cql(ip)
 ])
-success = run.run_pytest(sys.path[0], ['--host=' + ip] + sys.argv[1:])
+vnodes_success = run.run_pytest(sys.path[0], ['--host=' + ip, '--replication=vnodes'] + sys.argv[1:])
+tablets_success = run.run_pytest(sys.path[0], ['--host=' + ip, '--replication=tablets'] + sys.argv[1:])
```
Comment (nyh) [inline test/cql-pytest/run:31]: I don't like this change I don't think test/cql-pytest/run should run all the tests twice and have double the code for tracking passes and failures. I think it's complicates the code and confusing. I think the "run" script should run only one set of tests. If you want to run it once with tablets and once without, run the "run" script twice.

File: test/cql-pytest/run.py [hunk 4]
```diff
@@ -311,6 +311,8 @@ def run_scylla_cmd(pid, dir):
         # test/alternator/run.
         '--experimental-features=udf',
         '--experimental-features=keyspace-storage-options',
+        '--experimental-features=consistent-topology-changes',
+        '--experimental-features=tablets',
```
Comment (nyh) [inline test/cql-pytest/run.py:315]: This means test/cql-pytest/run will now **always** run Scylla with tablets enabled - just not used in the keyspace. I think this is fine at this stage of the development where consistent-topology-changes and tablets are stable "enough". We just need to be aware we did this. If you run with "vnode replication" it's not exactly the same as before this patch because of these two experimental flags.

**Review by nyh — CHANGES_REQUESTED** [review]

File: test/cql-pytest/conftest.py (hunk 1 above)
Comment (denesb) [reply test/cql-pytest/conftest.py:108]: Ok, so I'll leave only the vnodes one, and rename it to something else, to better express that this test is disabled temporarily, *until* the bug in the code causing it to fail is fixed. Everything has to support tablets, so something not working with tablets is a bug indeed.

File: test/cql-pytest/conftest.py (hunk 2 above)
Comment (denesb) [reply test/cql-pytest/conftest.py:102]: Good point, we don't want to stress-test scylla with 128 tablets in cql-pytest. It was a good experiment, and it shook out some bugs, but we shouldn't do this all the time. Your test adding-dropping tables is a more concentrated test for this.

**Comment by nyh** [comment]
> As expected running the tests with tablets fails. What is less expected, is that running with tablets takes twice as long: Vnodes:

Unfortunately, whereas in test/alternator I took extreme care to minimize the number of table creations/deletions and got to a point where 1000 tests run in about 20 seconds, in cql-pytest I (and other test authors) were less obsessive about test speed, and we have many many tests that create a new table just for the single test. I think this part is much slower with 128 tablets than it is with 16 vnodes. I'm hoping that if you go down from 128 tablets to 4, you'll get back the speed.

By the way, 270 seconds is already too much :-( One day I want to benchmark the tests (easy - run "pytest --durations=5") and find which tests are the slowest and fix them.

File: test/cql-pytest/run (hunk 3 above)
Comment (nyh) [reply test/cql-pytest/run:31]: What I'm suggesting is that you can have an option to run with tablets, and another to run without tablets - but if you want to run both you'll need to run it twice. Jenkins can be modified to run it twice. Or test.py can be modified to run the same suite twice, with two different parameters.

I'm worried that this is a slippery slope - today you added 2 runs, in 7 months we'll have 2^7 runs with 2^7 combinations of 7 options.

**Comment by denesb** [comment]
v2 (still RFC):
* Use automatic test parameterization, to run each test with tablets and vnodes respectively

File: test/cql-pytest/conftest.py (hunk 1 above)
Comment (denesb) [reply test/cql-pytest/conftest.py]: Ok, will do.

File: test/cql-pytest/run (hunk 3 above)
Comment (denesb) [reply test/cql-pytest/run]: Found a much more elegant solution, have a look at v2.

I don't think we will add any other feature, which will require running all tests with/without them, like we do with tablets right now.

File: test/cql-pytest/run.py (hunk 4 above)
Comment (denesb) [reply test/cql-pytest/run.py:315]: Yes, I think this should not cause a problem anymore. Especially not in a single-node test, this won't stress any of the new feature in consistent-topology-changes, so we are not exposed to any flakyness.

File: test/cql-pytest/run (hunk 3 above)
Comment (nyh) [reply test/cql-pytest/run]: We had **exactly** the same "with and without" need for schema changes over Raft, and had a "--raft" option in test/cql-pytest.
We'll have exactly the same thing if we ever want to rewrite a major feature - e.g:
1. LWT over raft instead of paxos
2. Strongly consistent tables
3. MV rewrite

**Comment by denesb** [comment]
v3:
* Add `xfail_tablets` and annotate all failing tests with it, together with an issue reference.

**Comment by denesb** [comment]
Tests are now all passing locally (except the xfailed ones of course).

**Comment by denesb** [comment]
@nyh if you think this PR is now acceptable, I will update the title and the cover-letter. Let's wait for a CI run in the meanwhile.

File: test/cql-pytest/conftest.py [hunk 5]
```diff
@@ -93,9 +93,14 @@ def this_dc(cql):
...
 @pytest.fixture(scope="session")
-def test_keyspace(cql, this_dc):
+def test_keyspace(request, cql, this_dc):
     name = unique_name()
-    cql.execute("CREATE KEYSPACE " + name + " WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 1 }")
+    replication_options = [("class", "NetworkTopologyStrategy"), (this_dc, 1)]
+    if request.param == "tablets":
+        replication_options.append(("initial_tablets", 4))
+        name += '_tablets'
+    serialized_options = ", ".join([f"'{key}' : '{value}'" for key, value in replication_options])
```
Comment (nyh) [inline test/cql-pytest/conftest.py:102]: I think you over-engineered this code with all the conversions ;-)
I'm thinking ahead and I think that very soon (maybe in a few days), this "initial_tablets" thing will no longer be the way to enable tablets, but this code with the format conversions will remain.
But this is a nitpick, it's not really a terrible problem ;-)

File: test/cql-pytest/conftest.py [hunk 6]
```diff
@@ -221,3 +226,20 @@ def temp_workdir():
...
+# Automatically parameterize each test, using test_keyspace (directly or
+# indirectly) with two test_keyspace instances: one which uses tablets, another
+# which uses vnodes.
+# The test_keyspace receives the automatic param, via request.param.
+def pytest_generate_tests(metafunc):
+    if 'test_keyspace' in metafunc.fixturenames:
+        metafunc.parametrize(
+                "test_keyspace",
+                ["tablets", "vnodes"],
+                indirect=True)
```
Comment (nyh) [inline test/cql-pytest/conftest.py:245]: This is a very cool trick. I still think it should NOT be the default to run all the tests twice (and as I said, a year from now it can become 2^N times), at least not without a special parameter. But since the code is now so NOT ugly, I'm less happy to fight against it.

But can you please at least give some thought to the question of when we'll REMOVE this double-testing code? Will we still continue to run each test twice even a year from now, when tablets will have been the default for a long time - but still some users still use vnodes so we can defend the need to continue testing it?

Please also note that in 99% (just a made-up statistic) of cql-pytest tests, tablets do not matter at all. Tablets do not change the correctness of CQL filtering, or TIMESTAMP formatting, or hundreds of other cql-pytest tests. I think we can even decide right now to test **only** with tablets, and stop testing with vnodes.

I wonder if we could we test everything with tablets except the xfail_tablet tests (whose number should go down hopefully soon), which will be tested without tablets?

File: test/cql-pytest/conftest.py (hunk 5 above)
Comment (denesb) [reply test/cql-pytest/conftest.py:102]: I think even in the post `initial_tablets` world, there will be something in the `WITH REPLICATION`, that will be used to mark the keyspace as using tablets. So maybe the "how" will change but the "where" won't (replication options).

File: test/cql-pytest/conftest.py (hunk 6 above)
Comment (denesb) [reply test/cql-pytest/conftest.py:245]: I think that functional tests (cql-pytest) and integration tests (dtest) will run with both with and without tablets until we retire vnode support for user tables. Until such time that we support users opting out of tablets, we cannot afford to have virtually no test coverage for them. AFAIK, we have 0 tests that specifically test that the vnodes algorithm works. Instead, it is tested indirectly by all the different functional and integration tests that happen to write and read from tables.

Comment (nyh) [reply test/cql-pytest/conftest.py:245]: > I think that functional tests (cql-pytest) and integration tests (dtest) will run with both with and without tablets until we retire vnode support for user tables.

Was the same true for "raft-based schema changes"? If not, why is this case different?

> Until such time that we support users opting out of tablets, we cannot afford to have virtually no test coverage for them. AFAIK, we have 0 tests that specifically test that the vnodes algorithm works. Instead, it is tested indirectly by all the different functional and integration tests that happen to write and read from tables.

It doesn't mean that developers need to run these tests twice every time they change the way that CQL filtering works. A different possibility is to add an option, and Jenkins but NOT most developers - will choose to run the tests twice.

Since I think in the context of the vnode-tablet differences it is actually dtests which are more important, you should first think what will be approach in dtest and then decide what this means for cql-pytest. If dtest will need to be run twice by Jenkins, then it can also do this for test.py. (and also note that you change only applies to cql-pytest - what about test/cql? test/alternator? test/toplogy? Why don't we need to run all of those in the same way?).

Personally, I think that test.py (and dtest) should have a tablets/vnode option, which affects the way they run Scylla (and should set a **default** initial_tablets and not require the test code to set it!), and Jenkins would run test.py and dtest twice - once with tablets and once with vnodes. The same option can be in `test/cql-pytest/run`, but I don't think it should run all tests twice.

**Comment by denesb** [comment]
v3.1:
* Updated existing comments and added some new ones.
* Updated PR title and description.

I think this PR is not ready to go in (provided we have a consensus of course).

**Comment by nyh** [comment]
By the way, https://github.com/scylladb/scylladb/pull/16364 if it goes in may also push us in the direction I recommended above. Yes, even then we **could** specify in code whether each test uses tablets or doesn't - but it is more natural to just run Scylla with tablets enabled by default for all keyspaces. It will also affect all suites that use scylla (including dtest) and won't need to be different for every suite.

**Comment by denesb** [comment]
> By the way, #16364 if it goes in may also push us in the direction I recommended above. Yes, even then we **could** specify in code whether each test uses tablets or doesn't - but it is more natural to just run Scylla with tablets enabled by default for all keyspaces. It will also affect all suites that use scylla (including dtest) and won't need to be different for every suite.

After we make tablets defaults, we will need code to opt out for the case where we want to use vnodes. So the code in `test_keyspaces` will be inverted: it will have code to add some option for the vnodes case.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping.

**Comment by denesb** [comment]
v4:
* Run only tablets when invoked via `run`, run all modes when invoked via `tests.py`

**Comment by denesb** [comment]
CI will fail. There is more work required to make this work, after https://github.com/scylladb/scylladb/pull/16364 went in.

File: test/cql-pytest/conftest.py (hunk 6 above)
Comment (nyh) [reply test/cql-pytest/conftest.py:256]: I still don't understand why we need the "--all" option, or the complications that come with it (even the renaming of the test keyspace isn't necessary without it). Why can't test.py just run the test suite twice? Won't you need this ability for other test suites as well? E.g., the C++ unit tests, how will you run those with and without tablets without just running them twice?

Comment (denesb) [reply test/cql-pytest/conftest.py:256]: I think running the suite twice would make the failure reporting and the investigation experience even worse then it is today. With the current approach, the name of the test tells you which variant failed -- vnodes or tablets. If we ran the suites twice, you would have to go hunting in the logs to find out which variant it is.

Also, I want to avoid touching the immensely complex `test.py` if possible. Any changes there will impact not just this, but other test suites too.

As for the other suites, we will see when we get there. I don't think we want to run the C++ suites with/without tablets. Those are unit tests (for the most part), which don't care what replication scylla uses. There are some functional tests in `test/boost/cql_query_test.cc`, we should just migrate those to `test/cql-pytest`, one day.

File: test/cql-pytest/run [hunk 7]
```diff
@@ -27,7 +27,16 @@ run.wait_for_services(pid, [
...
+
+args = ['--host=' + ip]
+if not any(map(lambda x: x.startswith('--replication'), sys.argv)):
+    # The default for --replication is "all". To faciliate iterative
+    # development, with short compile+test cycles, we only run the test with
+    # --replication=tablets, when running them via 'run'.
+    # If the user already provided this option, don't overwrite it.
+    args.append('--replication=tablets')
+
+success = run.run_pytest(sys.path[0], args + sys.argv[1:])
```
Comment (avikivity) [inline test/cql-pytest/run:39]: Can we keep `run` running a single selection of tablets/vnodes, and make the duplication happen in test.py? Perhaps as two virtual suites?

I don't know why I'm asking it, I don't have a concrete reason, it's just how I would have done it. That's not enough to justify a change.

**Comment by nyh** [comment]
> v4:
> 
>     * Run only tablets when invoked via `run`, run all modes when invoked via `tests.py`

I like this better, and have less reason to object now, but I am still not a fan of this whole approach. I still think the approach of cql-pytest being responsible for running each test twice is a mistake, and if you use this approach, you'll need to do the same kinds of tricks for all other test suites as well - textual CQL tests, Alternator tests, topology tests, ... I think the better approach to have have a Scylla configuration option whether it will use tablets or vnodes by default, and then run the test twice - once with option 1 and once with option 2.

Basically, it's the same thing as we have today with running the tests in debug build and release build - neither the C++ test, the Python tests, or the textual CQL tests or anything - know how to run themselves on two versions of Scylla. It's not their job.

File: test/cql-pytest/run (hunk 7 above)
Comment (nyh) [reply test/cql-pytest/run:39]: This is what I've also been asking. Just like test.py knows how to run cql-pytest twice, once for debug build and once for release build, and we don't have code in cql-pytest (or test/cql or test/alternator or test/boost or ...) that knows it should run each test twice in two build modes.

File: test/cql-pytest/run [hunk 8]
```diff
@@ -27,7 +27,16 @@ run.wait_for_services(pid, [
     lambda: run.check_rest_api(ip),
     lambda: check_cql(ip)
 ])
-success = run.run_pytest(sys.path[0], ['--host=' + ip] + sys.argv[1:])
+
+args = ['--host=' + ip]
+if not any(map(lambda x: x.startswith('--replication'), sys.argv)):
+    # The default for --replication is "all". To faciliate iterative
```
Comment (nyh) [inline test/cql-pytest/run:33]: Note that test/cql-pytest/run was supposed to be a shortcut - you can also run Scylla yourself and then "pytest", but the run script is easier.  After this patch, "run something.py::sometest"  does the good thing (run the test only once) but "pytest something.py::sometest" does what I don't like - run the requested test twice, once with tablets once without.

File: test/cql-pytest/conftest.py (hunk 6 above)
Comment (nyh) [reply test/cql-pytest/conftest.py:256]: I agree with Avi - as I noted above with a made-up statistic, 99% of the cql-pytest don't care about tablets vs vnodes - it's just one node, and then the test checks some small CQL feature if it returns the right thing or not. It will be very annoying and frustrating to see each of these tests run twice every single time. Note that I don't care if they run 2 or 6 or 74 times during the CI (today they already run between 3 and 3*3 times) - I just don't want developers to have to have to do it.

Comment (nyh) [reply test/cql-pytest/conftest.py:256]: And watch out not to use the fact that test.py is overly-complex as a reason to complicate cql-pytest. Yes, I agree with you that test.py is too complex, andand I've been saying this all along, but it doesn't mean I want to bring all that unnecessary complexity into cql-pytest.

Comment (denesb) [reply test/cql-pytest/conftest.py:256]: > Doesn't this apply to cql-pytest too? Does bug that cql-pytest tests is likely to be triggered by tablets?

Most of the tests in test/boost do not ever run a single query. They instantiate some object and call its API with different parameters and check the results. Tablets/vnodes have no bearing on these kinds of tests..
Like I said, the exception is `test/boost/cql_query_test.cc`, which *does* run queries.

Contrary to that, *all* tests in cql-pytest run queries, and those will all exercise the core replication paths.

File: test/cql-pytest/run (hunk 7 above)
Comment (denesb) [reply test/cql-pytest/run:39]: There is a very simple reason for building this switching into `run` + `conftest.py`: simplicity. I did this in less then 2 dozens of lines of code (half of which is comments).

If I were to do this in `test.py`, I'd had to write a dedicated test suite class + test class. The currently used PythonSuite + PythonTest are shared by other tests, so I can't just willy-nilly change it. Then I would have to make sure the test suites are distinguishable in the output, they match the command line filters, etc. This will be hundreds of lines of code and it would probably take me days to do.

> Just like test.py knows how to run cql-pytest twice, once for debug build and once for release build

This sounds simple but it is not. I would have to add a new global test parameter, like mode, which would result in changing **all** the existing test suites. Most test suites would not use this new parameter, but I would still have to make sure they don't choke on it. 

Maybe we'll have to do this anyway, when converting the other test suites. But I'm really not itching to do this before I know I must do it.

**Comment by denesb** [comment]
> Basically, it's the same thing as we have today with running the tests in debug build and release build - neither the C++ test, the Python tests, or the textual CQL tests or anything - know how to run themselves on two versions of Scylla. It's not their job.

All test suites will need to know about the two modes. For the simple reason, that soon, there will be no way to disable tablets on the command line. So tests will need to know to either disable or enable tablets when they create keyspaces and tables.

File: test/cql-pytest/run (hunk 8 above)
Comment (denesb) [reply test/cql-pytest/run:33]: > but "pytest something.py::sometest" does what I don't like - run the requested test twice, once with tablets once without

You can pass a command line param to `pytest` to disable this: `--replication=tablets`. I know you also don't like this.

**Comment by nyh** [comment]
> For the simple reason, that soon, there will be no way to disable tablets on the command line. So tests will need to know to either disable or enable tablets when they create keyspaces and tables.

This is not law of nature - we could, while making tablets default for new tables, also add an Scylla configuration option, say "--default-vnodes", which makes tablets **not** the default.

Making this a command line option (we'll need to add the same option to cql-repl and C++ tests as well) will make it easier to run our 20 different test suites - cql-pytest, cql, alternator, boost, and even dtest and SCT - with the default tablets as well as with vnodes (in separate Scylla runs) because you only need to change one place, where Scylla is configured, and don't need to hunt down dozens of uses of CREATE KEYSPACE. For example, dtest does not **really** have a single place where keyspaces are created - there's create_ks but a bunch of tests use CREATE KEYSPACE directly (https://github.com/scylladb/scylla-dtest/issues/3839).

**Comment by denesb** [comment]
I'm axing this PR and will send one with a much more reduced scope, of just enabling tablets by default in test.py and making sure tests pass, with minimal changes to cql-pytest itself.

I will leave figuring out how or if at all to run this suite against vnodes too. This sticky question however is holding up the entire effort of making tests work with tablets and by extension the tablet effort by large.

File: test/cql-pytest/conftest.py [hunk 9]
```diff
@@ -221,3 +235,22 @@ def temp_workdir():
     """ Creates a temporary work directory, for the scope of a single test. """
     with tempfile.TemporaryDirectory() as workdir:
         yield workdir
+
+@pytest.fixture(scope="function")
+def xfail_tablets(test_keyspace):
+    if test_keyspace.endswith('_tablets'):
+        pytest.xfail("test fails with tablets")
```
Comment (nyh) [inline test/cql-pytest/conftest.py:242]: this is the wrong way to do xfail - it forces an "XFAIL" status and doesn't even try to run the test (i.e., no chance for XPASS). I'll send my own PR in an hour, and you can see the right way - basically
```
request.node.add_marker(pytest.mark.xfail(reason='Test expected to fail with tablets experimental feature on'))
```

**Comment by denesb** [comment]
New PR: https://github.com/scylladb/scylladb/pull/16840

File: test/cql-pytest/conftest.py (hunk 9 above)
Comment (denesb) [reply test/cql-pytest/conftest.py:242]: No need to send yet another PR, I already have mine ready: https://github.com/scylladb/scylladb/pull/16840

**Comment by nyh** [comment]
> New PR: #16840

As I said, I'm also working on my own PR for the same thing because some of what I was proposing wasn't coming across. Then we can decide which one to take, or how to merge them (your original versions had some design errors that weren't even related to the idea of running each twice, I didn't review this new PR yet because I'm working on mine).

**Comment by denesb** [comment]
> > New PR: #16840
> 
> As I said, I'm also working on my own PR for the same thing because some of what I was proposing wasn't coming across. Then we can decide which one to take, or how to merge them (your original versions had some design errors that weren't even related to the idea of running each twice, I didn't review this new PR yet because I'm working on mine).

I'm fine with an alternative proposal, but let's try to minimize duplicate work. I had to adjust some tests to work with tablets, there is no point in you duplicating this effort. For this reason, maybe have a quick look at my new PR, before finishing yours. If after that, you still think an alternative proposal make sense, then finish and publish your own.
